python/ql/src/experimental/semmle/python/frameworks/Invoke.qll

```diff
@@ -0,0 +1,150 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `invoke` PyPI package.
+ * See https://www.pyinvoke.org/.
+ */
+
+private import python
+private import experimental.dataflow.DataFlow
+private import experimental.semmle.python.Concepts
+
+/**
+ * Provides models for the `invoke` PyPI package.
+ * See https://www.pyinvoke.org/.
+ */
+private module Invoke {
+  // ---------------------------------------------------------------------------
+  // invoke
+  // ---------------------------------------------------------------------------
+  /** Gets a reference to the `invoke` module. */
+  private DataFlow::Node invoke(DataFlow::TypeTracker t) {
+    t.start() and
+    result = DataFlow::importModule("invoke")
+    or
+    exists(DataFlow::TypeTracker t2 | result = invoke(t2).track(t2, t))
+  }
+
+  /** Gets a reference to the `invoke` module. */
+  DataFlow::Node invoke() { result = invoke(DataFlow::TypeTracker::end()) }
+
+  /**
+   * Gets a reference to the attribute `attr_name` of the `invoke` module.
+   * WARNING: Only holds for a few predefined attributes.
+   */
+  private DataFlow::Node invoke_attr(DataFlow::TypeTracker t, string attr_name) {
+    attr_name in ["run", "sudo", "context", "Context", "task"] and
+    (
+      t.start() and
+      result = DataFlow::importMember("invoke", attr_name)
+      or
+      t.startInAttr(attr_name) and
+      result = DataFlow::importModule("invoke")
+    )
+    or
+    // Due to bad performance when using normal setup with `invoke_attr(t2, attr_name).track(t2, t)`
+    // we have inlined that code and forced a join
+    exists(DataFlow::TypeTracker t2 |
+      exists(DataFlow::StepSummary summary |
+        invoke_attr_first_join(t2, attr_name, result, summary) and
+        t = t2.append(summary)
+      )
+    )
+  }
+
+  pragma[nomagic]
+  private predicate invoke_attr_first_join(
+    DataFlow::TypeTracker t2, string attr_name, DataFlow::Node res, DataFlow::StepSummary summary
+  ) {
+    DataFlow::StepSummary::step(invoke_attr(t2, attr_name), res, summary)
+  }
+
+  /**
+   * Gets a reference to the attribute `attr_name` of the `invoke` module.
+   * WARNING: Only holds for a few predefined attributes.
+   */
+  private DataFlow::Node invoke_attr(string attr_name) {
+    result = invoke_attr(DataFlow::TypeTracker::end(), attr_name)
+  }
+
+  /** Provides models for the `invoke` module. */
+  module invoke {
+    /** Gets a reference to the `invoke.context` module. */
+    DataFlow::Node context() { result = invoke_attr("context") }
+
+    /** Provides models for the `invoke.context` module */
+    module context {
+      /** Provides models for the `invoke.context.Context` class */
+      module Context {
```

Comment:
It feels like you could do with just one module. Do you expect to write predicates in `context` which are not in `Context` in the future?

Reply:
Agreed, we _could_. I was trying to keep the structure of the python code and the QL code very similar. I'm not 100% sure we want to keep everything private in the future, and if we expose it, I wanted to be prepared with a 1-1 correspondence.

This does pose the problem of where to put the model of a specific Python class. For example, the HTTP request class in django is defined in [`django.http.request.HttpRequest`](https://github.com/django/django/blob/b7da588e883e12b8ac3bb8a486e654e30fc1c6c8/django/http/request.py#L53), but the canonical way to use it in django code is to `from django.http import HttpRequest` -- so we need to figure out whether this should be modeled as `django::http::request::HttpRequest::class_()` or `django::http::HttpRequest::class_()` -- I'm leaning slightly towards mirroring the actual Python class location, but I'm also very open for suggestions and good arguments.

If you feel like this is a very bad idea, I'm open to discussing remove the `module context` :+1: (but here is at least the reason why I did it)

Reply:
It does seem appealing to mirror the Python class hierarchy...I guess for the http request class, we could define it where Python defines it and make an alias in the parent module..